jlm/mlir/backend/JlmToMlirConverter.cpp

```diff
@@ -0,0 +1,222 @@
+  }
+
+  // Create an MLIR operation for each RVSDG node and store each pair in a
+  // hash map for easy lookup of corresponding MLIR operation
+  std::unordered_map<rvsdg::node *, ::mlir::Value> nodes;
+  for (rvsdg::node * rvsdgNode : rvsdg::topdown_traverser(&region))
+  {
+    // TODO
+    // Get the inputs of the node
+    // for (size_t i=0; i < rvsdgNode->ninputs(); i++)
+    //{
+    //  ::llvm::outs() << rvsdgNode->input(i) << "\n";
+    //}
+    nodes[rvsdgNode] = ConvertNode(*rvsdgNode, block);
+  }
+
+  ::llvm::SmallVector<::mlir::Value> results;
+  for (size_t i = 0; i < region.nresults(); ++i)
+  {
+    auto output = region.result(i)->origin();
+    rvsdg::node * outputNode = rvsdg::node_output::node(output);
+    if (outputNode == nullptr)
+    {
+      // The result is connected directly to an argument
+      results.push_back(block.getArgument(output->index()));
+    }
+    else
+    {
+      // The identified node should always exist in the hash map of nodes
+      JLM_ASSERT(nodes.find(outputNode) != nodes.end());
+      results.push_back(nodes[outputNode]);
+    }
+  }
+  return results;
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertNode(const rvsdg::node & node, ::mlir::Block & block)
+{
+  if (auto simpleNode = dynamic_cast<const rvsdg::simple_node *>(&node))
+  {
+    return ConvertSimpleNode(*simpleNode, block);
+  }
+  else if (auto lambda = dynamic_cast<const llvm::lambda::node *>(&node))
+  {
+    return ConvertLambda(*lambda, block);
+  }
+  else
+  {
+    std::string unkown = "Unimplemented structural node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertSimpleNode(const rvsdg::simple_node & node, ::mlir::Block & block)
+{
+  if (auto bitsOp = dynamic_cast<const rvsdg::bitconstant_op *>(&(node.operation())))
+  {
+    auto value = bitsOp->value();
+    auto constOp = Builder_->create<::mlir::arith::ConstantIntOp>(
+        Builder_->getUnknownLoc(),
+        value.to_uint(),
+        value.nbits());
+    block.push_back(constOp);
+
+    return constOp;
+  }
+  else
+  {
+    std::string unkown = "Unimplemented simple node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertLambda(const llvm::lambda::node & lambdaNode, ::mlir::Block & block)
+{
+  ::llvm::SmallVector<::mlir::Type> arguments;
+  for (size_t i = 0; i < lambdaNode.nfctarguments(); ++i)
+  {
+    arguments.push_back(ConvertType(lambdaNode.fctargument(i)->type()));
+  }
+  ::llvm::ArrayRef argumentsArray(arguments);
+
+  ::llvm::SmallVector<::mlir::Type> results;
+  for (size_t i = 0; i < lambdaNode.nfctresults(); ++i)
+  {
+    results.push_back(ConvertType(lambdaNode.fctresult(i)->type()));
+  }
+  ::llvm::ArrayRef resultsArray(results);
+
+  ::llvm::SmallVector<::mlir::Type> lambdaRef;
+  auto refType = Builder_->getType<::mlir::rvsdg::LambdaRefType>(argumentsArray, resultsArray);
```

Comment:
Maybe you can inline argumentsArray and resultsArray?

`<...LambdaRefType>(::llvm::ArrayRef(argumentsArray), ::llvm::resultsArray(results))`

Reply:
Changed